### ISSUE TITLE
Fetch snapshot of previous version in read-write mode

### ISSUE THREAD

**Issue by nickasd** [share/sharedb#293]
I've implemented a search and replace feature and in this case it would be useful being able to apply the replace operations on the snapshot version at the time of the search; applying the replace operations on the current version could not work if a change was applied in the meantime.

It is currently possible to get a particular version of a document in read-only mode, but will it be possible to open it in read-write mode, such that any operation applied will be automatically merged with all the following versions?

**Comment by alecgibson** [comment]
We currently handle this by using `doc.pause()` when performing a find-and-replace operation. An alternative solution would be to cache the ops that have been received since doing the find-and-replace, so that they can be used to transform the replacement op (similar to what the PR for presence does, I believe).

**Comment by nickasd** [comment]
Thanks for your fast reply. Is this method documented somewhere?

**Comment by alecgibson** [comment]
Only if the [source code][1] counts as documentation. Was just looking through what we've done, and you'll need to couple this with `unsubscribe`:

  - `doc.pause()` will stop the document from _sending_ ops
  - `doc.unsubscribe()` will stop the document from _receiving_ ops

It might be nice if we had a method that would stop both (`doc.freeze()`?).

[1]: https://github.com/share/sharedb/blob/fe1b7c0549a9a1bda1de9d47a041426983ddd5c1/lib/client/doc.js#L824-L833

**Comment by alecgibson** [comment]
Actually yeah, looking through, it's definitely more important to `unsubscribe`. The `pause` is possibly not even necessary (especially if you do something like lock the UI).

**Comment by nickasd** [comment]
Ok, thanks. Actually I was looking for a way to apply an operation to a document that may have been modified in the meantime. This is the scenario:

1) I do a search operation.
2) Someone modifies the document.
3) I do a replace operation.

So with your suggested solution I would have to freeze the editor after step 1, unsubscribe every affected document, then unfreeze and subscribe again after step 3, right? (So no way of doing this while still receiving document updates.)

**Comment by alecgibson** [comment]
So the way we do this is:

1. I do a search (locally)
2. I listen for updates, and do another search when the document is updated
3. I start a replace operation
4. Document unsubscribes from further updates
5. We do another search now that the document is unsubscribed, to make sure we're completely up-to-date
6. Submit the replacement ops
7. Re-subscribe

**Comment by nickasd** [comment]
Do you think there's a way of doing this on the server? Because in my case there can be quite a lot of documents, which is why they can't all be downloaded to the client performing the replace operation.

**Comment by alecgibson** [comment]
Sure - performing operations on the server is really just like using another client, so anything you can do client-side, you can do server-side. It would probably look something like:

1. I submit a search (to the server)
2. Server responds with results
3. You do whatever it is you're doing to keep search results up-to-date when someone else changes the document
4. I submit a replace request
5. Server client unsubscribes itself from updates
6. Server client performs another search
7. Server client puts together a replacement op and submits
8. Server client resubscribes to updates / closes its connection / whatever cleanup you want

**Comment by nickasd** [comment]
I was just asking because there were some discrepancies between applying operations on the client/server before (I opened a ticket some time ago that was resolved), but I think that may not be the case anymore.

I was really looking for a way to not having to update the search results after the initial search and before the replace, because then the user would still have to review all the changes and still get some errors if the documents get changed between the client submitting the replace request and the server receiving and applying it.

**Comment by alecgibson** [comment]
I mean at this point it sort of becomes a matter of UX. As a user, if I issue the command "replace 'foo' with 'bar'", and I fire that off to the server, but get a document back that still contains 'foo' (which would happen if only applying on an historical document), I'd be kind of confused.

If this is really a path you want to follow, you can't edit an historical document (it doesn't really make any sense). The best thing you could do is probably something like:

1. Fetch snapshot at particular version/timestamp
2. Construct the op you'd need to apply for that version
3. Fetch most up-to-date version
4. Transform ops to apply to that version (eg by diffing the latest version and the historical version)
5. Submit the transformed ops

**Comment by nickasd** [comment]
I see your point and why the user could be confused. I was thinking the other way round, that the user could be confused if something was replaced that he did not confirm (because his version of the document was not up to date but the server applied the replace operation to the newer version). My approach would be of possibly replacing too little, and your approach would be of possibly replacing too much.

Perhaps in the end the best solution is to just try to replace exactly the text at the moment the search operation was done, with the downside that maybe a completely unrelated portion of the document keeps getting updated and hence the replace operation always fails.

I think I still prefer my solution, and after the replace operation another search is done and in case it still gives some results, the user can replace those too. What your solution wants to avoid (as I understand) is leaving the search term unreplaced in case it was inserted in the meantime, but it may still be the case that another user finishes typing that term shortly after the replace operation is done, or even starts typing it afterwards.

**Comment by alecgibson** [comment]
Sure it's definitely a trade-off. As I say, if you want to take you're approach, then the methodology I've outlined above should work.

**Comment by nickasd** [comment]
The difficult part remains this:
> Transform ops to apply to that version (eg by diffing the latest version and the historical version)
That's what I was looking for, not really a way of modifying a historical document, but rather a way of transforming an operation with respect to a historical document so that it would be applicable to the newest version of it. Is there anything I can use?

**Comment by alecgibson** [comment]
Here's an example of what you might do if you were using `rich-text`. You should be able to do something similar with other types: https://runkit.com/alecgibson/5ce26e8fc27749001aa5681b

**Comment by nickasd** [comment]
What is `rich-text`? I'm using JSON documents representing plain text files.

**Comment by nickasd** [comment]
Do you mean this [rich-text](https://www.npmjs.com/package/rich-text)? It doesn't seem to have received an update since 3 years.

Regardless, is it not something that should already be implemented somehow? When the server receives multiple operations at the same time, some of them are going to be applied to an older version of the document. I would say that this is exactly what I need in this case: transforming an operation with regard to an older version.

**Comment by alecgibson** [comment]
Yes, that's the one. It's what the Quill editor uses, although the underlying `Delta` class has been updated outside this package and this one's a bit out-of-date now (we're using a forked version).

Anyway, you can try to reconstruct an older version of the document and submit from there. Yes, in theory, it should get transformed properly. Probably something like:

```javascript
// Fetch the latest doc. We're going to replace its contents, but we need to
// set its connection, collection and ID. Could also just construct it.
const doc = connection.get('collection', 'id');

// Fetch, but don't subscribe to updates.
connection.fetchSnapshot('collection', 'id', oldVersionNumber, (error, snapshot) => {
  if (error) throw error;
  doc.ingestSnapshot(snapshot, (error) => {
    if (error) throw error;
    // Submit the replacement op constructed for the old version. The
    // backend should try to transform this to be appropriate to the current
    // version.
    doc.submitOp(replacementOp, (error) => {
      if (error) throw error;
      // Subscribe to get latest updates
      doc.subscribe((error) => {
        if (error) throw error;
      });
    });
  });
});
```

**Comment by nickasd** [comment]
Thanks for your help. Is this what the server does already when handling multiple concurrent operations? Because if yes, then it would make sense for me to make a function out of this, so that I (and possibly other developers) could just use it without worrying that it won't work anymore in the future.

**Comment by alecgibson** [comment]
In terms of how the server handles multiple ops, that's all happening behind-the-scenes in `submitOp`, which is what this code snippet is leaning on.

Note that the above is just how I think it should all roughly be put together, and I haven't tested it, so please do have a go and report back. And by all means if you think it's useful for other devs, then wrapping it in a utility method could be useful.

**Comment by nickasd** [comment]
Well, I was hoping that someone who knows the code better than me ;-) could extract that part into a function with signature like `applyHistoricalOp(document, version, op)`, do you see what I mean? I would assume that the server could use it when receiving an operation from a client applied to a specific version.

**Comment by alecgibson** [comment]
We've discussed this at the weekly Pull Request meeting, and we don't think it's particularly idiomatic to apply ops to an historical document version. We'd recommend one of:

- Unsubscribe from the document when performing the search, and resubscribing after submitting the replacement
- Transforming the op to apply to the latest version (like my `rich-text` example)

**Comment by nickasd** [comment]
I think you misunderstood me, or I didn't explain the right way. I don't want to apply operations to a historical document, but to transform them with regard to a historical document. The function above should rather be named `transformHistoricalOp(document, version, op)`: it should transform the operation `op`, which is based on the state of `document` with `version`, so that it can then be applied to the current version of that document. This way a user could perform a search, and even if the relevant document is edited in the meantime, the replace operation could be transformed with the above function to be now relative to the most recent version of the document, and the replace operation would succeed regardless of what was changed in the meantime.

**Comment by alecgibson** [comment]
What's wrong with just unsubscribing from the document?

**Comment by nickasd** [comment]
I don't understand why you suggest unsubscribing now, because what I want is your second suggestion:
> Transforming the op to apply to the latest version (like my rich-text example)

Or do you see it differently?

The replace operation is done on the server since the document collection can be quite large, so the client would have to download it all first, which should be avoided in this case. The server would then have to unsubscribe from the documents which should be kept around indefinitely since it doesn't know when the client will perform the replace operation, if at all.

**Comment by alecgibson** [comment]
If you want to transform the op, then things obviously get a bit trickier, especially if you want to just use standard methods defined on OT types. Here are some approaches.

## Transform ops yourself using a delta

The most "logical" approach is probably to diff the two documents, and then transform your replacement op to apply to the latest version, as I outlined above for `rich-text`. Here's the same example adapted for `json0`: https://runkit.com/alecgibson/5ce5822561339f001a496872

Note that this needs non-standard methods for diffing, which requires these extra modules:

 - [`json0-ot-diff`](https://www.npmjs.com/package/json0-ot-diff)
 - [`diff-match-patch`](https://www.npmjs.com/package/diff-match-patch)

## Transform ops one-by-one

The way ShareDB handles the case of old version is by fetching all the ops between the two documents, and transforming them one-by-one. This avoids the step calculating the Delta, but obviously requires that you fetch all the ops.

## Lean on ShareDB

The final approach is to lean on the internals of ShareDB that already do this, as I [outlined above](https://github.com/share/sharedb/issues/293#issuecomment-493928241). Here's an example of that in action: https://runkit.com/alecgibson/5ce656247e042f001a051117

**Comment by nickasd** [comment]
I had a closer look at your sample (the one leaning on ShareDB) and I realized that it does in fact do what I want, so sorry for the confusion. I guess what confused me is that you suggested this "fetch snapshot/submit op/subscribe" sequence, while in fact I expected to see a function that would transform an operation with respect to a historical document version.

While playing around with this implementation, I noticed an odd behaviour. If I have a document `{title: b}`, then execute these two parallel operations `[{ p: ['title', 0], si: 'a' }, { p: ['title', 2], si: 'c' }]` and `[{ p: ['title', 0], sd: 'b' }, { p: ['title', 0], si: 'B' }]`, I get an error `Delete component 'b' does not match deleted text 'a'`. See the code [here](https://runkit.com/nickasd/5cef04366e2c4b001c2a3e1f). Can you explain this?

**Comment by alecgibson** [comment]
The problem is that Share isn't really designed for (actively) modifying older versions of documents. There's a bunch of stuff here that leads to this behaviour:

 - When you do `connection.get`, ShareDB gives you a cached document (if available), so your documents are actually both the same object
 - When you submit an op on an old version of a document, it will also fetch all the newer ops and apply them, so that it's up-to-date
 - Considering both the above, then you probably get into some sort of weird race condition where you have the same instance of a document attempting to submit two ops at a particular version, but after one has been submitted, it transforms itself up to the most recent version, which probably then doesn't play nicely when attempting to submit on old on the same document again.
 - You can avoid this behaviour by using different connections in parallel, which can have different instances of the document that don't (directly) affect one another

**tl;dr**: use different connections for each document. See fixed example [here](https://runkit.com/alecgibson/5cef82349a7e000020c4b796)

**Comment by nickasd** [comment]
Thank you as always for your fast response.
> When you submit an op on an old version of a document, it will also fetch all the newer ops and apply them, so that it's up-to-date

I don’t get this. What’s the use of `doc.subscribe()` then?

**Comment by alecgibson** [comment]
> I don’t get this. What’s the use of doc.subscribe() then?

The `doc.subscribe` ensures that you get any new ops _after_ you're up-to-date (ie you _stay_ up-to-date).

> ...the second operation produces the text Bac. Why is that?

I'm not very familiar with the `text0` type. I suspect this happens because `text0` has no "replace" operation (cf. `json0`'s ability to do `{p: ['path'], oi: {foo:true}, od: {bar:false}}`).

It might become a bit clearer if you ignore your deletion, and consider this case of two parallel ops:

```javascript
[{ p: ['title', 0], si: 'a' }, { p: ['title', 2], si: 'c' }]
```

```javascript
[{ p: ['title', 0], si: 'B' }]
```

Either one of these may go first, so you could end up with either `aBbc` (if the top op goes first) or `Babc` (if the bottom op goes first).

By the time we acknowledge that either of these cases can happen, then we can clearly see that if we _now_ delete `b`, then we end up with either `aBc` (which is what you expected), or `Bac`, which is the other case.

This is indeed a little odd and counter-intuitive, but I believe that the behaviour is correct, given that there is no "replace" in `text0`.

**Comment by nickasd** [comment]
> When you submit an op on an old version of a document, it will also fetch all the newer ops and apply them, so that it's up-to-date

I see. Perhaps this could be mentioned in the README, because I don't think this is an intuitive behaviour. I would have expected that a document remains "offline" until I subscribe it again.

> Either one of these may go first, so you could end up with either aBbc (if the top op goes first) or Babc (if the bottom op goes first).

Perhaps the use of the term "parallel" was misleading here. What I would expect, since both operations are applied to the same document version, is that the order of applying them is irrelevant, so I should end up with `aBc` in any case. The two operations run in parallel in some sense, but the latter should be implicitly transformed with regard to the first one. Do you see what I mean? Isn't that what applying operations to a historical version and then subscribing again should do?

**Comment by alecgibson** [comment]
I think what you want here isn't ShareDB's use-case. ShareDB is designed to keep a single, canonical version of a document that multiple clients can edit simultaneously. The fact that it can apply ops to an old version is a necessary by-product of this primary focus.

I'm not sure I understand why you should always end up with `aBc`. As I explained above: these ops may be applied to the single, canonical version of the document (ie the version in the data store) in either order, and hence you may end up with either `aBc` or `Bac`, depending on which op hits the data store first.

**Comment by nickasd** [comment]
I though this was a possible use case: user X and Y make two edits which arrive at the server at the same time, for example the two operations above (one wants to make the `b` uppercase and the other one wants to prepend an `a` and append a `c`). User X would expect to end up with `B` and given Y's change ends up with `aBc`. User Y would expect to end up with `abc` and given X's change ends up with `aBc` as well. If this is not what ShareDB is supposed to do, then I think I completely missed the point.

**Comment by alecgibson** [comment]
I understand what the intent is, but this is unfortunately one of those times where this OT implementation just fails to capture that intent. Consider a string `abc` and this simple op:

```javascript
[{p: ['myString', 1], si: 'd'}]
```

Now, given this representation, it's impossible for ShareDB to know exactly what I mean. I could mean any of:

 1. insert 'd' after 'a'
 2. insert 'd' before 'b'
 3. insert 'd' at exactly position `1` (regardless of what other changes have been made around it)
 4. insert 'd' at the position that is two characters before the end of the string

Obviously all of these intents would lead to slightly different merges when a conflicting edit arrives. For example, if another user transforms the string to `xaybc`, then the above intents would all want slightly different behaviour:

  1. `xadybc`
  2. `xaydbc`
  3. `xdaybc`
  4. `xaydbc`

What you're running into here is the fact that ShareDB makes some merge conflict choices for you, but merge conflicts (as any Git user should know) can be non-trivial for exactly this reason: Git can't infer intent, and neither can ShareDB. So it makes a good guess and leaves it up to the users to fix any weird cases that arise from the (unlikely, but certainly not impossible) event of a weird merge.

To quote that wiki article, "Unfortunately, implementing OT sucks."

**Comment by nickasd** [comment]
> What you're running into here is the fact that ShareDB makes some merge conflict choices for you

It would be interesting now to know what choices are made, if it can be explained in a few sentences. I didn't realize that in this simple example such a choice has to be made, since I would have expected it to work this way:
1. Assume the initial string `b`, now X and Y send concurrent operations.
2. Edit Y arrives first: `[{ p: ['title', 0], si: 'a' }, { p: ['title', 2], si: 'c' }]` changes the string to `abc`.
3. Edit X arrives last: `[{ p: ['title', 0], sd: 'b' }, { p: ['title', 0], si: 'B' }]`. It should delete the character `b` at position 0 and insert `B` at position 0. Since operation Y inserted a character at position 0 before, all following operations at position >=0 are shifted by +1. The same would apply for the insertion of `c` at position 2, but in the example there are no following edits with position >=2.

In the Wikipedia article I see that (in some cases) positions are shifted only if they are strictly > than the position of a previous insert operation, instead of the >= I had assumed. Is this how it's done in ShareDB? Do you think that > is more intuitive than >=? As seen in the example, using >= instead of > would support a correct replace operation; are there scenarios where using > instead of >= behaves better?

**Comment by alecgibson** [comment]
My own opinion is that - given the ambiguity - there's always going to be users who expect it to act one way, and other users who expect otherwise, but honestly, that level of detail a bit beyond my understanding of ShareDB and `json0`/`text0`. You'll want to dig around in the [`json0`][1] repo.

ShareDB is put together in such a way that you're free to fork `json0` and use your own version, or - if you're super keen - you can create your own OT type. For example, if this was genuinely something you wanted to pursue, you can conceivably clear up this sort of ambiguity with an OT type that has special disambiguation methods, eg `replace`, `insertBefore`, `insertAfter`, etc., but that's very much left as an exercise to the reader.

[1]: https://github.com/ottypes/json0

**Comment by nickasd** [comment]
Ok, thank you for the input. I will keep using the existing implementation for now.

Regarding the `fetchSnapshot` method, I'm still not sure if I'm using it correctly. As you can see in [this example](https://runkit.com/nickasd/5cf18f07b30b24001eeba529), I get two different documents when loading the same version twice.

**Comment by alecgibson** [comment]
As I've pointed out before, every time you use `submitOp`, ShareDB will fetch the latest version of the document as part of its OT process.

**Comment by nickasd** [comment]
That’s true, but I don’t see why this is relevant in this case. Immediately after the second call to `ingestSnapshot` the document title is already wrong, i.e. it is `xabc` when version 1 is `abc`.

**Comment by alecgibson** [comment]
You're not using `doc.destroy()` correctly; it's asynchronous, and you should execute further code in a callback. Also you will then need to re-fetch the document.

https://runkit.com/alecgibson/5cf3a34509671b001d5ef387

Any further issues should probably be raised in a new issue, because this isn't really related to the original issue any more.

**Comment by nickasd** [comment]
Thanks, using `doc.destroy()` in the correct places solved my problem. Explaining this in the documentation would be very welcome.

**Comment by NakedFool** [comment]
> 
> 
> As I've pointed out before, every time you use `submitOp`, ShareDB will fetch the latest version of the document as part of its OT process.

I think i know where my problem is then if submitOp is taking most recent version.

Let me explain, i'm using fetchSnapshot so user can jump from version to version, however when i edit old version it does not become last one, i'm assuming since as you said submitOp will always fetch most recent version.

For example 
Most recent version : 'LastVersion'
Previous: 'PreviousVersion' and if i return to this version and add some characters for instance it becomes 'PreviousVersion**1**' 
Most recent version: 'LastVersion**1**' instead of 'PreviousVersion1'

Do you have any idea how can i let subscription know to add previous version on  top instead of just adding some characters i typed on top of the most recent version?

**Comment by alecgibson** [comment]
@NakedFool I'm not sure I follow. Can you please open a new issue with a minimum working example of your problem?

**Comment by NakedFool** [comment]
@alecgibson 
Will do!
